README.md
<p align="center">

  <a href="https://github.com/zeeshanayaz/custom_accordion/">
    <img src="https://img.shields.io/badge/platforms-iOS%20%7C%20macOS%20%7C%20Android%20%7C%20Linux%20%7C%20Windows%20%7C%20Web-green.svg" alt="Supported platforms" />
  </a>
  <br>
  <a href="https://pub.dartlang.org/packages/custom_accordion">
    <img src="https://img.shields.io/pub/v/custom_accordion.svg" alt="Pub package version." />
  </a>
  <a href="https://opensource.org/licenses/MIT">
    <img src="https://badges.frapsoft.com/os/v1/open-source.svg?v=102" alt="Open Source Love." />
  </a>
  <a href="https://github.com/zeeshanayaz/custom_accordion/blob/master/LICENSE">
    <img src="https://img.shields.io/badge/license-MIT-blue.svg" alt="Custom Accordion is released under the MIT license." />
  </a>
</p>

# Custom Accordion

Create dynamic, interactive, and visually appealing accordions in your Flutter app with ease using the "Custom Accordion" package. This powerful library simplifies the process of adding collapsible sections to your UI, making it a breeze to manage and customize accordions to suit your design and functionality needs.

## Table of contents 
- [Demo Preview](#demo-preview)
- [Get Started](#get-started)
- [How to use](#how-to-use)
- [Properties](#properties)
- [Contribution](#contribution)


## Demo Preview

<div style="display: flex; justify-content: center;" >
  <img src="https://github.com/zeeshanayaz/custom_accordion/raw/master/demo/custom_accordion_example.gif" style="width: 50%;"/>
</div>

## Get Started

### Add to your project
```yaml
dependencies:
  custom_accordion: ^version
```

or simply run the command in your terminal at project root directory

```
flutter pub add custom_accordion
```

### Import and use it!
```
import 'package:custom_accordion/custom_accordion.dart';
```

## How to use

Basic Accordion
```dart
    CustomAccordion(
        title: 'Accordion',
        ),
```

Accordion with no child
```dart
    CustomAccordion(
        title: 'Accordion with no child',
        subTitle: 'Accordion Sub Title',
        headerBackgroundColor: Colors.blue,
        titleStyle: const TextStyle(
          fontSize: 18,
          fontWeight: FontWeight.bold,
          color: Colors.white,
        ),
        subTitleStyle: const TextStyle(
          fontSize: 16,
          fontWeight: FontWeight.normal,
          color: Colors.white,
        ),
        toggleIconOpen: Icons.keyboard_arrow_down_sharp,
        toggleIconClose: Icons.keyboard_arrow_up_sharp,
        headerIconColor: Colors.white,
        accordionElevation: 0,
        showContent: true,
        ),
```

Accordion with one child
```dart
    CustomAccordion(
        title: 'Accordion with one child',
        subTitle: 'Accordion Sub Title',
        headerBackgroundColor: Colors.blue,
        titleStyle: const TextStyle(
          fontSize: 18,
          fontWeight: FontWeight.bold,
          color: Colors.white,
        ),
          subTitleStyle: const TextStyle(
          fontSize: 16,
          fontWeight: FontWeight.normal,
          color: Colors.white,
        ),
        toggleIconOpen: Icons.keyboard_arrow_down_sharp,
        toggleIconClose: Icons.keyboard_arrow_up_sharp,
        headerIconColor: Colors.white,
        accordionElevation: 0,
        widgetItems: const Text(
          'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.',
          ),
        ),
```
Accordion with many child
```dart
    CustomAccordion(
        title: 'Accordion with many child',
        headerBackgroundColor: Colors.blue,
        titleStyle: const TextStyle(
          fontSize: 18,
          fontWeight: FontWeight.bold,
          color: Colors.white,
        ),
        toggleIconOpen: Icons.keyboard_arrow_down_sharp,
        toggleIconClose: Icons.keyboard_arrow_up_sharp,
        headerIconColor: Colors.white,
        accordionElevation: 0,
        widgetItems: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              Text(
                'Lorem Ipsum',
                style: Theme.of(context).textTheme.titleMedium,
              ),
              ElevatedButton(
                onPressed: () {},
                child: const Text(
                  'ELEVATED BUTTON',
                ),
              )
            ],
          ),
        ),
```

Accordion with List Items and disabled toggle example
```dart
        CustomAccordion(
            title: 'Accordion with List Items',
            subTitle: 'Accordion toggle disabled',
            headerBackgroundColor: Colors.blue,
            titleStyle: const TextStyle(
              fontSize: 18,
              fontWeight: FontWeight.bold,
              color: Colors.white,
            ),
            subTitleStyle: const TextStyle(
              fontSize: 16,
              fontWeight: FontWeight.normal,
              color: Colors.white,
            ),
            toggleIconOpen: Icons.keyboard_arrow_down_sharp,
            toggleIconClose: Icons.keyboard_arrow_up_sharp,
            headerIconColor: Colors.white,
            accordionElevation: 0,
            disableToggle: true,
            widgetItems: Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: List<Widget>.generate(5, (int index) {
                return Text('Item $index');
                }),
            ),
        ),
```

Please see example project for code.


## Properties

The following properties can be set for the widget.

| Property              | Description                                                     | Type               | Required | Default value                          |
|-----------------------|-----------------------------------------------------------------|--------------------|----------|----------------------------------------|
| title                 | Title of Accordion                                              | String             | Yes      |                                        |
| subTitle              | Sub Title of Accordion                                          | String             | No       |                                        |
| headerBackgroundColor | Header background color                                         | Color              | No       | Theme.of(context).splashColor          |
| titleStyle            | Header text style                                               | TextStyle          | No       | Theme.of(context).textTheme.titleLarge |
| subTitleStyle         | Sub Title text style                                            | TextStyle          | No       | Theme.of(context).textTheme.titleSmall |
| toggleIconOpen        | Toggle icon to expand/open widgets                              | IconData           | No       | Icons.arrow_drop_down                  |
| toggleIconClose       | Toggle icon to collapse/close widgets                           | IconData           | No       | Icons.arrow_drop_up                    |
| headerIconColor       | Icon Color                                                      | Color              | No       | Theme.of(context).iconTheme.color      |
| backgroundColor       | Background color of widget                                      | Color              | No       | Theme.of(context).cardColor            |
| accordionElevation    | Elevation of accordion widget                                   | double             | No       | Theme.of(context).cardTheme.elevation  |
| widgetItems           | Child Widget(s) of accordion                                    | Widget             | No       | Text()                                 |
| widgetItemsPadding    | Padding arround children of accordion                           | EdgeInsetsGeometry | No       | EdgeInsets.all(10)                     |
| showContent           | Make default behaviour of accordion children visible or hidden  | bool               | No       | false                                  |
| disableToggle         | Make default behaviour of accordion toggle                      | bool               | No       | false                                  |


## Contribution
🚀 We believe in the power of collaboration and welcome contributions from our amazing community to make **Custom Accordion** even more awesome!

Whether you're a developer, designer, or just an enthusiast, you can contribute in various ways:

* 💡 Feature Ideas: Have a cool feature in mind? Share your ideas with us. We love fresh perspectives!
* 🐛 Bug Reports: Spotted a bug? Let us know! Help us squash those pesky little critters.
* 📖 Documentation: Improve our documentation. Clear and concise docs make a world of difference.
* 🛠 Code Contributions: If you're a developer, dive into the codebase. Tackle open issues or implement new features.
* 💬 Community Support: Share your knowledge! Help others by answering questions and engaging in discussions.
* ✏️ Design: If you have design skills, contribute by enhancing our UI/UX.

Dig into [**CONTRIBUTING.MD**](CONTRIBUTING.md), which covers submitting bugs, requesting new features, preparing your code for a pull request, etc.

Please don't forget to **like**, **follow**, and **star our repo**!.

## Support the package (optional)
If you find this package useful, you can support it for free by giving it a **thumbs up** at the top of this page.

## Bugs or Requests
If you encounter any problems feel free to open an [issue](https://github.com/zeeshanayaz/custom_accordion/issues/new?template=bug_report.md). If you feel the library is missing a feature, please raise a [ticket](https://github.com/zeeshanayaz/custom_accordion/issues/new?template=feature_request.md). Pull request are also welcome.

## License
This package is available under the [**MIT License**](https://github.com/zeeshanayaz/custom_accordion/blob/master/LICENSE).


CHANGELOG.md
## 1.0.0
* Documentation updated

## 0.0.5
* Documentation updated

## 0.0.4
* Documentation updated

## 0.0.3

* Added sub title under accordion heading
* handled toggle functionality as disableToggle

## 0.0.2
* Demo preview added in documentation.

## 0.0.1
* initial release.


test/custom_accordion_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:custom_accordion/custom_accordion.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/custom_accordion.dart
library custom_accordion;

import 'package:flutter/material.dart';

// ignore: must_be_immutable
class CustomAccordion extends StatefulWidget {
  CustomAccordion({
    super.key,
    required this.title,
    this.subTitle,
    this.headerBackgroundColor,
    this.titleStyle,
    this.subTitleStyle,
    this.toggleIconOpen,
    this.toggleIconClose,
    this.headerIconColor,
    this.backgroundColor,
    this.accordionElevation,
    this.widgetItems,
    this.widgetItemsPadding,
    this.showContent = false,
    this.disableToggle = false,
  });

  /// title: Title heading of Accordion
  /// required
  final String title;

  /// subTitle: Subtitle
  /// Optional property
  final String? subTitle;

  /// headerBackgroundColor visible at the back of title
  /// Optional property
  /// default: Theme.of(context).splashColor
  final Color? headerBackgroundColor;

  /// titleStyle: Text style of title
  /// Optional property
  /// default: Theme.of(context).textTheme.titleLarge
  final TextStyle? titleStyle;

  /// subTitleStyle: Text style of subTitle
  /// Optional property
  /// default: Theme.of(context).textTheme.titleSmall
  final TextStyle? subTitleStyle;

  /// toggleIconOpen: Icon displayed when the accordion is in collapsed state
  /// Optional property
  /// default: Icons.arrow_drop_down
  final IconData? toggleIconOpen;

  /// toggleIconClose: Icon displayed when the accordion is in expanded state
  /// Optional property
  /// default: Icons.arrow_drop_up
  final IconData? toggleIconClose;

  /// headerIconColor: Toggle Icon color
  /// Optional property
  /// default: Theme.of(context).iconTheme.color
  final Color? headerIconColor;

  /// backgroundColor: Background color of body of accordion
  /// Optional property
  /// default: Theme.of(context).cardColor
  final Color? backgroundColor;

  /// accordionElevation: Elevation of Accordion
  /// Optional property
  /// default: Theme.of(context).cardTheme.elevation
  final double? accordionElevation;

  /// widgetItems: Widget(s)
  /// Optional property
  /// default:
  ///               Center(
  ///                  child: Text(
  ///                    'No Widget available',
  ///                    style: Theme.of(context).textTheme.bodyMedium,
  ///                  ))
  final Widget? widgetItems;

  /// widgetItemsPadding: Padding around body of Accordion
  /// Optional property
  /// default: const EdgeInsets.all(10)
  final EdgeInsetsGeometry? widgetItemsPadding;

  /// showContent: default state of accordion toggle
  /// if 'true', default body visibility will be visible
  /// Optional property
  /// default: false
  bool showContent;

  /// disableToggle: Control the toggle functionality.
  /// if 'true', You won't see toggle icons and accordion won't toggle its state
  /// Optional property
  /// default: false
  bool disableToggle;

  @override
  State<CustomAccordion> createState() => _CustomAccordionState();
}

class _CustomAccordionState extends State<CustomAccordion> {
  @override
  Widget build(BuildContext context) {
    return Card(
      clipBehavior: Clip.hardEdge,
      elevation:
          widget.accordionElevation ?? Theme.of(context).cardTheme.elevation,
      color: widget.backgroundColor ?? Theme.of(context).cardColor,
      child: Column(children: [
        ListTile(
          onTap: widget.disableToggle ? null : () {
            setState(() {
              widget.showContent = !(widget.showContent);
            });
          },
          tileColor:
              widget.headerBackgroundColor ?? Theme.of(context).splashColor,
          title: Text(
            widget.title.toUpperCase(),
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
            style: widget.titleStyle ??
                Theme.of(context).textTheme.titleLarge,
          ),
          subtitle: widget.subTitle != null ? Text(
            widget.subTitle ?? '',
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
            style: widget.subTitleStyle ??
                Theme.of(context).textTheme.titleSmall,
          ) : null,
          trailing: widget.disableToggle ? null : IconButton(
            icon: Icon(
              widget.showContent
                  ? (widget.toggleIconClose ?? Icons.arrow_drop_up)
                  : (widget.toggleIconOpen ?? Icons.arrow_drop_down),
              color:
              widget.headerIconColor ?? Theme.of(context).iconTheme.color,
            ),
            tooltip: 'Toggle Accordion',
            splashRadius: 20,
            onPressed: () {
              setState(() {
                widget.showContent = !(widget.showContent);
              });
            },
          ),
        ),
        widget.disableToggle
          ? Container(
          padding: widget.widgetItemsPadding ?? const EdgeInsets.all(10),
          child: widget.widgetItems ??
              Center(
                  child: Text(
                    'No Widget available',
                    style: Theme.of(context).textTheme.bodyMedium,
                  )),
        )
          : (widget.showContent)
            ? Container(
                padding: widget.widgetItemsPadding ?? const EdgeInsets.all(10),
                child: widget.widgetItems ??
                    Center(
                        child: Text(
                      'No Widget available',
                      style: Theme.of(context).textTheme.bodyMedium,
                    )),
              )
            : const SizedBox(),
      ]),
    );
  }
}


example/lib/main.dart
import 'package:custom_accordion/custom_accordion.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Custom Accordion Demo',
      theme: ThemeData(),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({super.key});

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.yellow,
      appBar: AppBar(
        title: const Text('Custom Accordion Demo'),
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            /// Basic Accordion
            CustomAccordion(
              title: 'Accordion',
            ),

            /// Accordion with no child
            CustomAccordion(
              title: 'Accordion with no child',
              subTitle: 'Accordion Sub Title',
              headerBackgroundColor: Colors.blue,
              titleStyle: const TextStyle(
                fontSize: 18,
                fontWeight: FontWeight.bold,
                color: Colors.white,
              ),
              subTitleStyle: const TextStyle(
                fontSize: 16,
                fontWeight: FontWeight.normal,
                color: Colors.white,
              ),
              toggleIconOpen: Icons.keyboard_arrow_down_sharp,
              toggleIconClose: Icons.keyboard_arrow_up_sharp,
              headerIconColor: Colors.white,
              accordionElevation: 0,
              showContent: true,
            ),

            /// Accordion with one child
            CustomAccordion(
              title: 'Accordion with one child',
              subTitle: 'Accordion Sub Title',
              headerBackgroundColor: Colors.blue,
              titleStyle: const TextStyle(
                fontSize: 18,
                fontWeight: FontWeight.bold,
                color: Colors.white,
              ),
              subTitleStyle: const TextStyle(
                fontSize: 16,
                fontWeight: FontWeight.normal,
                color: Colors.white,
              ),
              toggleIconOpen: Icons.keyboard_arrow_down_sharp,
              toggleIconClose: Icons.keyboard_arrow_up_sharp,
              headerIconColor: Colors.white,
              accordionElevation: 0,
              widgetItems: const Text(
                'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.',
              ),
            ),

            /// Accordion with many children
            CustomAccordion(
              title: 'Accordion with many children',
              headerBackgroundColor: Colors.blue,
              titleStyle: const TextStyle(
                fontSize: 18,
                fontWeight: FontWeight.bold,
                color: Colors.white,
              ),
              toggleIconOpen: Icons.keyboard_arrow_down_sharp,
              toggleIconClose: Icons.keyboard_arrow_up_sharp,
              headerIconColor: Colors.white,
              accordionElevation: 0,
              widgetItems: Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: [
                  Text(
                    'Lorem Ipsum',
                    style: Theme.of(context).textTheme.titleMedium,
                  ),
                  ElevatedButton(
                    onPressed: () {},
                    child: const Text(
                      'ELEVATED BUTTON',
                    ),
                  )
                ],
              ),
            ),

            /// Accordion with List Items and Toggle Disabled
            CustomAccordion(
              title: 'Accordion with List Items',
              subTitle: 'Accordion toggle disabled',
              headerBackgroundColor: Colors.blue,
              titleStyle: const TextStyle(
                fontSize: 18,
                fontWeight: FontWeight.bold,
                color: Colors.white,
              ),
              subTitleStyle: const TextStyle(
                fontSize: 16,
                fontWeight: FontWeight.normal,
                color: Colors.white,
              ),
              toggleIconOpen: Icons.keyboard_arrow_down_sharp,
              toggleIconClose: Icons.keyboard_arrow_up_sharp,
              headerIconColor: Colors.white,
              accordionElevation: 0,
              disableToggle: true,
              widgetItems: Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: List<Widget>.generate(5, (int index) {
                  return Text('Item $index');
                }),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
